Draft day4 2 (#25)
UW-Madison-DataScience/geospatial-capstone: _episodes_rmd/05-day4.Rmd

---
title: 'Day 4: Exploring Vector Data'
exercises: 7
questions: 
- "Are two datasets from two different sources suitable to be plotted together?" 
- "What spatial patterns can we see by visual analysis?"  
- "What are some plotting options to impact the way data are visualized?" 
output: html_document
objectives:
- "Importing vector data and visually explore the vector data through attributes and by visualization."
keypoints: 
 - "exploring vector data attributes"
 - "consider design options for your vector data"

>  - "visual analysis of plotted data"

Reviewer: For these key points, it may be helpful to phrase them as meaningful things for them to remember. I adapted and selected some of the key points from the lessons. It was helpful to use what existed so I didn't feel like I had to make up something new. Here's a lesson that may be relevant for some key points (at the bottom): https://uw-madison-datascience.github.io/r-raster-vector-geospatial/06-vector-open-shapefile-in-r/index.html

## 1) Load Libraries 
Load these libraries if you don't already have them (`dplyr`, `ggplot2`, `sf`, `rgdal`)

> ## Solution
> ```{r, eval=TRUE}
> library(dplyr)
> library(ggplot2)
> library(sf)
> library(rgdal)
> ```
{: .solution}

## 2) Load vector files
In this exercise, we have provided two capstone vector data files: Wisconsin Annual Precipitation Average 1981-2010, and county outlines. You can work with these, or with your own vector datasets.

Filenames: precip1981_2010_a_wi.shp  and  WI_Counties2010.shp

> ## Solution
> ```{r, eval=TRUE}
> wi_precip <- st_read("../data/precipitation/precip1981_2010_a_wi.shp")
> wi_county <- st_read("../data/WI_Counties2010/WI_Counties2010.shp")
> ## County data is from here  https://geo.btaa.org/catalog/4E677AF3-3FF2-43FA-8A58-3C345EC7F465
```
{: .solution}

## 3) Explore metadata

> We have worked with commands to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 

Reviewer: I have often used the term "functions" rather than "commands".

> We have worked with commands to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 

Author: ```suggestion
We have worked with functions to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 
```

> We have worked with commands to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 

Author: ```suggestion
We have worked with functions to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 
```

> Explore the metadata of the shapefiles. We will work towards comnining these files in the same plot. Are these files suited to being plotted together - for example, are they in the same CRS and cover the same area? 

Reviewer: ```suggestion
Explore the metadata of the shapefiles. We will work towards combining these files in the same plot. Are these files suited to being plotted together - for example, are they in the same CRS and cover the same area? 
```

> ## Solution
> ```{r, eval=TRUE}
> st_geometry_type(wi_precip)
> st_crs(wi_precip)$proj4string
> st_bbox(wi_precip)

> st_geometry_type(wi_county)
> st_crs(wi_county)$proj4string
> st_bbox(wi_county)

```
> The CRS for both vectors is the same. Both are in longlat projection, datum WGS84.

> > You may notice that the bounding box limits are slightly different. Since the limits are similar, we will continue working with them with the intent to plot them together, and make a note to inspect the plots and see why the limits are different. 

Reviewer: I love this part! It's great to highlight that they should inspect their data and mentally flag a potential issue, but just because they aren't the same doesn't mean it won't work.

{: .solution}

## 4) Explore attributes. Identify attribute(s) of interest

> If you are working with our capstone data, identify the attribute that appears to contain the precipitation data, and the attribute of county name. If you are working with your own data, choose an attribute of interest to work with for plotting.

Reviewer: ```suggestion
If you are working with the capstone data, identify the attribute that appears to contain the precipitation data, and the attribute of county name. If you are working with your data, choose an attribute of interest to work with for plotting.
```

There are multiple ways to explore the attributes of a spatial data file. We have worked with the `names()` function and previewing the first 6 rows with the `head()` functions as well as viewing the file in RStudio's Environment tab, or printing the shapefile object to the screen.

For a visual method to explore attributes, you can try the `plot()` function built into sf. This plots a small map for each of the attributes. For example 
`plot(wi_precip)`
Notice that this `plot()` command displays attribute names but does not display data values.
```{r}
plot(wi_precip)
plot(wi_county)
```
{: .solution}

## 5) Using ggplot2 to plot data with selected attribute of interest

> If working with our capstone data,  plot the precipitation shapefile using the atribute you identified as the precipitation data to set the fill. One way to do this is to use a fill aesthetic mapped to the column name of the attribute you identified `aes(fill = )`  

Reviewer: ```suggestion
If working with the capstone data,  plot the precipitation shapefile using the attribute you identified as the precipitation data to set the fill. One way to do this is to use a fill aesthetic mapped to the column name of the attribute you identified `aes(fill = )`  
```

>   geom_sf(data = wi_county, fill = NA, color = gray(.5)) +  
>   scale_fill_continuous(trans = 'reverse') +        ## reverses color scale
>   ggtitle("Wisconsin Annual Precipitation Average 1981 - 2010, In Inches") +
>   coord_sf()
> ```
{: .solution}

## Bonus: Try different color palette
R offers a number of color packages that allow you to choose from a variety of color ranges. One example of a color palette is  `colorspace` 

The `hcl_palettes` function will print the colorspace palette to the screen

 ```{r, eval=TRUE}
 # install.packages("colorspace")
 library(colorspace)
 hcl_palettes(plot = TRUE)
 ```

You can assign colors from a palette to the unique values in a range, as in our precipitation data. First, determine the number of unique values.

 ```{r, eval=TRUE}
 wi_precip_uniq <- unique(wi_precip$Inches)
 length(wi_precip_uniq)
 ```

In the case of our precipitation data, there are 11 unique values. This code example selects the palette "Purples 3" and reverses the order of the colors.

```{r, eval=TRUE}
colors <- sequential_hcl(11, palette = "Purples 3")  #11 categories
colors <- rev(colors)   #reverses order
```

> Commands to plot the map are similiar to previous examples, with the addition of the `cut_interval()` function which creates groups with equal ranges of values.

Reviewer: The cut_intervals() function is a little confusing to me. I think it would be helpful to add an explanation here about why this is necessary, and a longer description about what the function does.

---
title: 'Day 4: Exploring Vector Data'
exercises: 7
questions: 
- "Are two datasets from two different sources suitable to be plotted together?" 

> - "What can spatial patterns can we see by visual analysis?"  

Reviewer 2: ```suggestion
- "What spatial patterns can we see by visual analysis?"  
```

- "What are some plotting options to impact the way data are visualized?" 
output: html_document
objectives:

> - "Importing vector data; exploring vector data through attributes and by plotting"

Reviewer 2: ```suggestion
- "Importing vector data and visually explore the vector data through attributes and by visualization."
```

keypoints: 
 - "exploring vector data attributes"
 - "consider design options for your vector data"
 - "visual analysis of plotted data"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## Uncomment out the next two lines when committing
source("../bin/chunk-options.R")
knitr_fig_path("05-")
```

In this capstone lesson, we will work with two Wisconsin vector data files: precipitation, and county boundaries. If you are using your own data, please identify 2 different vector files that you can use to summarize and visualize in this lesson. The learning objectives of this capstone are to demonstrate an example workflow of loading, exploring, and plotting vectors. 

## 1) Load Libraries 

> Load these libraries if you don't already have them (dplyr, ggplot2, sf, rgdal)

Reviewer 2: ```suggestion
Load these libraries if you don't already have them (`dplyr`, `ggplot2`, `sf`, `rgdal`)
```

> ## Solution
> ```{r, eval=TRUE}
> library(dplyr)
> library(ggplot2)
> library(sf)
> library(rgdal)
> ```
{: .solution}

## 2) Load vector files
In this exercise, we have provided two capstone vector data files: Wisconsin Annual Precipitation Average 1981-2010, and county outlines. You can work with these, or with your own vector datasets.

Filenames: precip1981_2010_a_wi.shp  and  WI_Counties2010.shp

> ## Solution
> ```{r, eval=TRUE}
> wi_precip <- st_read("../data/precipitation/precip1981_2010_a_wi.shp")
> wi_county <- st_read("../data/WI_Counties2010/WI_Counties2010.shp")
> ## County data is from here  https://geo.btaa.org/catalog/4E677AF3-3FF2-43FA-8A58-3C345EC7F465
```
{: .solution}

## 3) Explore metadata

> We have worked with commands to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type`, `st_crs`,  `st_bbox`, as well as reviewing the output produced in R when you imported the shapefiles. 

Reviewer 2: ```suggestion
We have worked with commands to explore the geometry type, and projection of vector data files. Examples include:`st_geometry_type()`, `st_crs()`,  `st_bbox()`, as well as reviewing the output produced in R when you imported the shapefiles. 
```

Explore the metadata of the shapefiles. We will work towards comnining these files in the same plot. Are these files suited to being plotted together - for example, are they in the same CRS and cover the same area? 

> ## Solution
> ```{r, eval=TRUE}
> st_geometry_type(wi_precip)
> st_crs(wi_precip)$proj4string
> st_bbox(wi_precip)

> st_geometry_type(wi_county)
> st_crs(wi_county)$proj4string
> st_bbox(wi_county)

```
> The CRS for both vectors is the same. Both are in longlat projection, datum WGS84.
> You may notice that the bounding box limits are slightly different. Since the limits are similar, we will continue working with them with the intent to plot them together, and make a note to inspect the plots and see why the limits are different. 
{: .solution}

## 4) Explore attributes. Identify attribute(s) of interest
If you are working with our capstone data, identify the attribute that appears to contain the precipitation data, and the attribute of county name. If you are working with your own data, choose an attribute of interest to work with for plotting.

> There are multiple ways to explore the attributes of a spatial data file. We have worked with the `names` function and previewing the first 6 rows with the `head` functions as well as viewing the file in RStudio's Environment tab, or printing the shapefile object to the screen.

Reviewer 2: ```suggestion
There are multiple ways to explore the attributes of a spatial data file. We have worked with the `names()` function and previewing the first 6 rows with the `head()` functions as well as viewing the file in RStudio's Environment tab, or printing the shapefile object to the screen.
```

> For a visual method to explore attributes, you can try the `plot` function built into sf. This plots a small map for each of the attributes. For example 

Reviewer 2: ```suggestion
For a visual method to explore attributes, you can try the `plot()` function built into sf. This plots a small map for each of the attributes. For example 
```

`plot(wi_precip)`

> Notice that this `plot` command displays attribute names but does not display data values.

Reviewer 2: ```suggestion
Notice that this `plot()` command displays attribute names but does not display data values.
```

```{r}
plot(wi_precip)
plot(wi_county)
```
{: .solution}

## 5) Using ggplot2 to plot data with selected attribute of interest
If working with our capstone data,  plot the precipitation shapefile using the atribute you identified as the precipitation data to set the fill. One way to do this is to use a fill aesthetic mapped to the column name of the attribute you identified `aes(fill = )`  

> ## Solution
> ```{r, eval=TRUE}
> # Using aes in  ggplot2 
> ggplot() +
>  geom_sf(data = wi_precip, aes(fill = PrecipInch), size = 0.1) +
>  ggtitle("Wisconsin Annual Precipitation Average 1981 - 2010, In Inches") +
>  coord_sf()
```
{: .solution}

> ## 6) Spatial Questions Discussion 

Reviewer 2: These questions are awesome!

Author: Thanks!

What do you see in the plotted map? For example, where are the highest/lowest values, and are there patterns across the state? How many different categories are there?
Notice the default colors and how they are associated with the data values. Do you think this color scale is a good choice for the data?
Can you easily read the different color values on the plot, and connect them to the data value? Is the color range intuitive to understand? Could this be symbolized better?

It is a common convention in map symbology to use darker colors for higher data values. We can do this in ggplot2 by reversing the color range. 
Add the line `scale_fill_continuous(trans = 'reverse') +`

> ## Solution
> ```{r, eval=TRUE}
> #Plot a shapefile, reverse the colors
> ggplot() +
>  geom_sf(data = wi_precip, aes(fill = PrecipInch), size = 0.1) +
>  scale_fill_continuous(trans = 'reverse') +   ## reverses color scale     
>  ggtitle("Wisconsin Annual Precipitation Average 1981 - 2010, In Inches") +
>  coord_sf()
```
{: .solution}

## 7) Plot multiple vector shapefiles

> You can use ggplot2 to plot mutiple vector shapefiles layered in a single plot. For example, adding a layer of the county outlines over the plot of precipitation with a second use of the `geom_sf` command. One design option to try is to make choices about the color and thickness of the county lines.

Reviewer 2: ```suggestion
You can use ggplot2 to plot mutiple vector shapefiles layered in a single plot. For example, adding a layer of the county outlines over the plot of precipitation with a second use of the `geom_sf()` command. One design option to try is to make choices about the color and thickness of the county lines.
```

>   geom_sf(data = wi_county, fill = NA, color = gray(.5)) +  
>   scale_fill_continuous(trans = 'reverse') +        ## reverses color scale
>   ggtitle("Wisconsin Annual Precipitation Average 1981 - 2010, In Inches") +
>   coord_sf()
> ```
{: .solution}

## Bonus: Try different color palette
R offers a number of color packages that allow you to choose from a variety of color ranges. One example of a color palette is  `colorspace` 

The `hcl_palettes` function will print the colorspace palette to the screen

 ```{r, eval=TRUE}
 # install.packages("colorspace")
 library(colorspace)
 hcl_palettes(plot = TRUE)
 ```

You can assign colors from a palette to the unique values in a range, as in our precipitation data. First, determine the number of unique values.

 ```{r, eval=TRUE}
 wi_precip_uniq <- unique(wi_precip$Inches)
 length(wi_precip_uniq)
 ```

In the case of our precipitation data, there are 11 unique values. This code example selects the palette "Purples 3" and reverses the order of the colors.

```{r, eval=TRUE}
colors <- sequential_hcl(11, palette = "Purples 3")  #11 categories
colors <- rev(colors)   #reverses order
```

> Commands to plot the map are similiar to previous examples, with the addition of the `cut_interval` function which creates groups with equal ranges of values.

Reviewer 2: ```suggestion
Commands to plot the map are similiar to previous examples, with the addition of the `cut_interval()` function which creates groups with equal ranges of values.
```
